sapi/fpm/fpm/fpm_status.c

```diff
@@ -45,6 +46,97 @@ int fpm_status_init_child(struct fpm_worker_pool_s *wp) /* {{{ */
 }
 /* }}} */
 
+/* {{{ proto array fastcgi_get_status
+ * Returns the status of the fastcgi process manager */
+PHP_FUNCTION(fastcgi_get_status)
+{
+	struct fpm_scoreboard_s scoreboard, *scoreboard_p;
+	struct fpm_scoreboard_proc_s proc;
+	zval fpm_proc_stats, fpm_proc_stat;
+	time_t now_epoch;
+	struct timeval duration, now;
+	double cpu;
+	int i;
+
+
+	scoreboard_p = fpm_scoreboard_get();
+	if (!scoreboard_p) {
+		zlog(ZLOG_ERROR, "status: unable to find or access status shared memory");
+		return;
+	}
+
+	if (!fpm_spinlock(&scoreboard_p->lock, 1)) {
+		zlog(ZLOG_NOTICE, "[pool %s] status: scoreboard already in used.", scoreboard_p->pool);
+		return;
+	}
+	/* copy the scoreboard not to bother other processes */
+	scoreboard = *scoreboard_p;
+	fpm_unlock(scoreboard_p->lock);
+
+	now_epoch = time(NULL);
+	fpm_clock_get(&now);
+
+	array_init(return_value);
+	add_assoc_string(return_value, "pool", scoreboard.pool);
+	add_assoc_string(return_value, "process-manager", PM2STR(scoreboard.pm));
+	add_assoc_long(return_value, "start-time", scoreboard.start_epoch);
+	add_assoc_long(return_value, "start-since", now_epoch - scoreboard.start_epoch);
+	add_assoc_long(return_value, "accepted-conn", scoreboard.requests);
+#ifdef HAVE_FPM_LQ
+	add_assoc_long(return_value, "listen-queue", scoreboard.lq);
+	add_assoc_long(return_value, "max-listen-queue", scoreboard.lq_max);
+	add_assoc_long(return_value, "listen-queue-len", scoreboard.lq_len);
+#endif
+	add_assoc_long(return_value, "idle-processes", scoreboard.idle);
+	add_assoc_long(return_value, "active-processes", scoreboard.active);
+	add_assoc_long(return_value, "total-processes", scoreboard.idle + scoreboard.active);
+	add_assoc_long(return_value, "max-active-processes", scoreboard.active_max);
+	add_assoc_long(return_value, "max-children-reached", scoreboard.max_children_reached);
+	add_assoc_long(return_value, "slow-requests", scoreboard.slow_rq);
+
+	array_init(&fpm_proc_stats);
+	for(i=0; i<scoreboard.nprocs; i++) {
+		if (!scoreboard_p->procs[i] || !scoreboard_p->procs[i]->used) {
+			continue;
+		}
+		proc = *scoreboard_p->procs[i];
```

Comment:
I'm wondering if this is actually safe to do without the lock as technically some value could be changed by other process during this copy. I'm aware that this is also done in `fpm_status_handle_request` for full syntax but not sure if it's correct in there too.

Reply:
meaning it might be a bug in there too... :)

Reply:
It struck me too when copy pasting it. I've been reading through the code over the past days.
I assume the worst possible result right now is an inconsistent reading, like number of active processes not being in sync with detailed process stats.
I assume the safe way to do it would be to lock the scoreboard, then iterate through the procs and copy (lock copy unlock) them one by one and only then unlocking the scoreboard.
Is this the path of action you're suggesting?

Reply:
Yeah I think that would be safe as we can't just copy the whole scoreboard and procs in one go (each proc can be locked individually..).

It would be good to introduce a new function `fpm_scoreboard_copy` that would allocate new scoreboard and copy all data into it (first locked scoreboard data and then iterate and copy the procs as you suggested). I think this might make sense to do in separate PR and use it for full syntax. The reason is that this part could be back ported to the stable branches as it seems like a bug to me. This PR can be then based on it.

Reply:
I opened https://github.com/php/php-src/pull/3185 and implemented safe access in 9eb8db6cea3e5aad0047eca7a8dba4dd67e0caaa